worlds/oot_soh/location_access/overworld/thieves_hideout.py

```diff
@@ -3,7 +3,233 @@
+                                         bundle) and world.options.fortress_carpenters.value == 1))
+                         and has_item(LocalEvents.TH_DOUBLE_CELL_COULD_FREE_DOUBLE_CELL_CARPENTER, bundle)
+                         and has_item(LocalEvents.TH_STEEP_SLOPE_CELL_COULD_FREE_SLOPE_CARPENTER, bundle)
+                         and has_item(LocalEvents.TH_DEAD_END_CELL_COULD_FREE_DEAD_END_CARPENTER, bundle)
+                         and has_item(LocalEvents.TH_1_TORCH_CELL_COULD_FREE_1_TORCH_CARPENTER, bundle))
+                        or world.options.fortress_carpenters.value == 2)
+    ])
+
+    ##Thieves Hideout Steep Slope Cell
+    # Events
+    add_events(Regions.THIEVES_HIDEOUT_STEEP_SLOPE_CELL, world, [
+        (EventsLocations.TH_STEEP_SLOPE_CELL_FREE_SLOPE_CARPENTER, LocalEvents.TH_STEEP_SLOPE_CELL_COULD_FREE_SLOPE_CARPENTER,
+         lambda bundle: can_kill_enemy(bundle, Enemies.GERUDO_WARRIOR))
+    ])
+    # Locations
+    add_locations(Regions.THIEVES_HIDEOUT_STEEP_SLOPE_CELL, world, [
+        (Locations.TH_STEEP_SLOPE_CARPENTER, lambda bundle: can_kill_enemy(bundle, Enemies.GERUDO_WARRIOR)),
+        (Locations.TH_STEEP_SLOPE_CELL_RIGHT_POT, lambda bundle: can_break_pots(bundle)),
+        (Locations.TH_STEEP_SLOPE_CELL_LEFT_POT, lambda bundle: can_break_pots(bundle))
+    ])
+    # Connections
+    connect_regions(Regions.THIEVES_HIDEOUT_STEEP_SLOPE_CELL, world, [
+        (Regions.GF_ABOVE_GTG, lambda bundle: True),
+        (Regions.GF_TOP_OF_LOWER_VINES, lambda bundle: True),
+        (Regions.THIEVES_HIDEOUT_RESCUE_CARPENTERS,
+         lambda bundle: (((small_keys(Items.GERUDO_FORTRESS_SMALL_KEY, 4,
+                                      bundle) and world.options.fortress_carpenters.value == 0)
+                          or (small_keys(Items.GERUDO_FORTRESS_SMALL_KEY, 1,
+                                         bundle) and world.options.fortress_carpenters.value == 1))
+                         and has_item(LocalEvents.TH_DOUBLE_CELL_COULD_FREE_DOUBLE_CELL_CARPENTER, bundle)
+                         and has_item(LocalEvents.TH_STEEP_SLOPE_CELL_COULD_FREE_SLOPE_CARPENTER, bundle)
+                         and has_item(LocalEvents.TH_DEAD_END_CELL_COULD_FREE_DEAD_END_CARPENTER, bundle)
+                         and has_item(LocalEvents.TH_1_TORCH_CELL_COULD_FREE_1_TORCH_CARPENTER, bundle))
+                        or world.options.fortress_carpenters.value == 2)
+    ])
+
+    ##TH Rescue Carpenters
+    ### This is a deviation from ship logic due to the union of locations
+    # Events
+    add_events(Regions.THIEVES_HIDEOUT_RESCUE_CARPENTERS, world, [
+        (EventsLocations.TH_RESCUED_ALL_CARPENTERS, Events.RESCUED_ALL_CARPENTERS,
+         lambda bundle: True)
+    ])
+    # Locations
+    add_locations(Regions.THIEVES_HIDEOUT_RESCUE_CARPENTERS, world, [
+        (Locations.GF_GERUDO_MEMBERSHIP_CARD,
+         lambda bundle: True)
+    ])
```

Comment:
And then this can check for `has_item(Events.RESCUED_ALL_CARPENTERS, bundle)` instead